kube/src/api/subresource.rs

```diff
@@ -208,3 +210,134 @@ where
         Ok(self.client.request_text_stream(req).await?)
     }
 }
+
+// ----------------------------------------------------------------------------
+// Attach subresource
+// ----------------------------------------------------------------------------
+/// Params for attaching
+#[derive(Default)]
+pub struct AttachParams<R, W>
+where
+    R: AsyncRead + Unpin,
+    W: AsyncWrite + Unpin,
+{
+    /// The container for which to attach. Defaults to only container if there is one container in the pod.
+    pub container: Option<String>,
+    /// If provided, the standard out stream of the pod will be redirected to it.
+    pub stdout: Option<W>,
+    /// If provided, the standard error stream of the pod will be redirected to it.
+    pub stderr: Option<W>,
+    /// If provided, the contents will be redirected to the standard input stream of the pod.
+    pub stdin: Option<R>,
+    /// Allocate a terminal for this attach call; defaults to `false`.
+    pub tty: bool,
+}
+
+impl Resource {
+    /// Attach to a pod
+    pub fn attach<R, W>(&self, name: &str, ap: &AttachParams<R, W>) -> Result<http::Request<()>>
+    where
+        R: AsyncRead + Unpin,
+        W: AsyncWrite + Unpin,
+    {
+        let base_url = self.make_url() + "/" + name + "/" + "attach?";
+        let mut qp = url::form_urlencoded::Serializer::new(base_url);
+
+        if ap.stdin.is_some() {
+            qp.append_pair("stdin", "true");
+        }
+        if ap.stdout.is_some() {
+            qp.append_pair("stdout", "true");
+        }
+        if ap.stderr.is_some() {
+            qp.append_pair("stderr", "true");
+        }
+        if ap.tty {
+            qp.append_pair("tty", "true");
+        }
+        if let Some(container) = &ap.container {
+            qp.append_pair("container", &container);
+        }
+
+        let req = http::Request::post(qp.finish());
+        req.body(()).map_err(Error::HttpError)
+    }
+}
+
+/// Marker trait for objects that has attach
+pub trait AttachingObject {}
+
+impl AttachingObject for k8s_openapi::api::core::v1::Pod {}
+
+impl<K> Api<K>
+where
+    K: Clone + DeserializeOwned + AttachingObject,
+{
+    /// Attach to pod
+    pub async fn attach<R, W>(&self, name: &str, mut ap: AttachParams<R, W>) -> Result<()>
+    where
+        R: AsyncRead + Unpin,
+        W: AsyncWrite + Unpin,
+    {
+        let req = self.resource.attach(name, &ap)?;
+        let stream = self.client.connect(req).await?;
```

Comment:
ah interesting, all websocket methods is an infinite loop after a connect. Guess that makes sense. Have you thought about how this thing could be cancelled, like signal handling?

Reply:
In general, `await` points also act as cancellation points (since they can return control upwards in the control stack and make the `Future` `Drop`pable).

Reply:
ah, that's true. I guess I was thinking a little bit about graceful shutdowns, but that probably isn't a big deal in the attach case